[계산기] 이승엽  최종제출합니다.

SEUNGBOONG opened the pull request:
사칙 연산 프로그램 최종 제출합니다 .
- 최종적으로 끝낼 것으로 예상 하고 있습니다.
- Controller에서 작동하지 않은 이슈 해결했습니다. 
- 처음에 이슈 내용에서는 Stack--force 를 확인한 결과 praseInput에서 문제라는 것을 보고 그부분만 수정을 진행했었습니다.
- 수정 후 작동해보니 정상적으로 작동한는 것을 보고
- 객체를 잘나누고 진행을 하면 수정이 편이하다는 것을 몸소 깨달았습니다. 감사합니다.
- OuputView에서도 출력값만을 확인할 수 있게 진행했는데 맞는지는 확실하진 않은 것 같습니다 .

SEUNGBOONG committed:
docs: 기능 구현 목록

SEUNGBOONG committed:
docs: 사칙연산 테스트 완성

SEUNGBOONG committed:
docs: 사칙연산 테스트 완성

SEUNGBOONG committed:
docs : 사칙연산 메서드 구현

SEUNGBOONG committed:
docs : 사칙연산 메서드 구현

SEUNGBOONG committed:
docs : Step1-3 README.md 작성 연습

SEUNGBOONG committed:
refactor :
MVC 모델로 재구성하기 위한 첫 번째 과정

SEUNGBOONG committed:
refactor : README.md 요구사항에 맞춰 리팩토링 진행했습니다.

SEUNGBOONG committed:
refactor : Arithmetic
Calculatr.java에 맞춰 테스트를 진행했습니다.

SEUNGBOONG committed:
refactor :조건에는 없지만, Controller에서 혹시 필요할까 넣었습니다.

SEUNGBOONG committed:
refactor : 리드미 재정비했습니다.

SEUNGBOONG committed:
refactor : domain에서
구분짓는 메서드를 위한 클래스를 새로 생성해서 구현했습니다. 그리고 요구사항에 있는 예외처리를 이용해봤는데 맞는지는 의구심이듭니다.

SEUNGBOONG committed:
refactor : 예외 처리 및 ArithmeticCalculatrTest 역시 여기서 진행했는데 이렇게 해도 되는지는 잘모르겠습니다.

SEUNGBOONG committed:
refactor : MVC 모델을 지켜가며 구현해봤습니다.

sosow0212 commented on README.md:
```diff
@@ -1,3 +1,18 @@
 
 계산기 미션 저장소
+
+## 기능 구현 목록
```
기능 구현 목록 좋습니다 👍

SEUNGBOONG replied:
감사합니다 !!

sosow0212 commented on src/main/java/controller/CalculatorController.java:
```diff
@@ -0,0 +1,21 @@
+        ac.plusCalculate();
+        inputview.printing();
+    }
+
+
```
불필요한 공백이 있습니다!

SEUNGBOONG replied:
감사합니다!

sosow0212 commented on src/main/java/controller/CalculatorController.java:
```diff
@@ -0,0 +1,21 @@
+public class CalculatorController {
+
+    public void controller() {
+
```
불필요한 공백 있습니다~

SEUNGBOONG replied:
신경쓰겠습니다.

sosow0212 commented on src/main/java/controller/CalculatorController.java:
```diff
@@ -0,0 +1,21 @@
+
+    public void controller() {
+
+        SplitString ss = new SplitString();
+        ArithmeticCalculator ac = new ArithmeticCalculator();
```
1. `ss`, `ac`는 축약어라 어떤 것을 의미하는지 읽기 힘드네요 ㅠㅠ 풀어서 써주거나 뜻을 부여해주면 더 좋을 것 같아요
2. 생성자로 객체를 주입해주는 것이랑, 현재 방식처럼 직접 `= new ...()`로 선언해주는 건 어떤 차이가 있을까요?

SEUNGBOONG replied:
필요한 부분에 컨벤션을 지키는 것이 아직 눈에띄지 않았습니다. 
수정하겠습니다!

sosow0212 commented on src/main/java/domain/SplitString.java:
```diff
@@ -0,0 +1,30 @@
+
+public class SplitString {
+
+    private final String DISTINCTION = ":|@|,";
```
1. 해당 정규표현식은 어떤 것을 의미하나요?
2. 상수는 `static final`로 선언해주세요!
3. 문제 요구사항이랑 조금 다른 것 같습니다! `//;\n1;2;3` 이렇게 입력값이 오는 경우에도 잘 작동하나요?

SEUNGBOONG replied:
문제 요구사항을 잘못파악했습니다.  ; @ , 이아니라 // \n 사이에 문자로 하는 것이었습니다. 수정하겠습니다. 
혹시 String으로 데이터를 받았을 때,  "//" 이부분과 \n 이부분은 인덱싱해서 분류해서 해야할까요?

sosow0212 replied:
필요하다면 Split으로 분리해서 원하는 값만 뽑아서 쓰면 될 것 같습니다!

sosow0212 commented on src/main/java/domain/SplitString.java:
```diff
@@ -0,0 +1,30 @@
+public class SplitString {
+
+    private final String DISTINCTION = ":|@|,";
+    private int i;
```
i는 굳이 클래스 필드로 가지고 있지 않아도 될 것 같습니다.
밑에 for에서 사용하는데 `for(int i=... )`로 for 내부에서 선언해도 될 것 같아요!

sosow0212 commented on src/main/java/domain/SplitString.java:
```diff
@@ -0,0 +1,30 @@
+
+    private final String DISTINCTION = ":|@|,";
+    private int i;
+    private final int INITIAL_NUMBER = 0;
```
이것도 static final로 수정해주시고 상수끼리 선언 해주세요!
선언 순서도 컨벤션 한 번 확인해주세요~

sosow0212 commented on src/main/java/domain/SplitString.java:
```diff
@@ -0,0 +1,30 @@
+    public int[] intArr;
+
+    public void splitStr(String str) {
+        String[] strArr = str.split(DISTINCTION);
```
문자열을 정규표현식으로 분리하는 것인가요? 의도한대로 작동되지 않을 것 같아요! 확인 부탁드릴게요

sosow0212 commented on src/main/java/domain/SplitString.java:
```diff
@@ -0,0 +1,30 @@
+        for (i = INITIAL_NUMBER; i < strArr.length; i++) {
+            try {
+                intArr[i] = Integer.parseInt(strArr[i]);
+                if (intArr[i] < 0) {
+                    throw new IllegalArgumentException();
+                }
```
1. 메서드 명이 Exception로 대문자로 시작하는데 동사 + 소문자로 시작해주세요!
2. 0보다 낮은 경우는 계산이 되지 않을 것 같은데 `나눗셈인 경우만 나누는 수가 0이 안되어야할 것`같아요!
3. intArr와 같은 네이밍은 지양해주세요

sosow0212 commented on src/main/java/domain/SplitString.java:
```diff
@@ -0,0 +1,30 @@
+        }
+    }
+
+
```
불필요한 공백 제거해주세욥

sosow0212 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,7 @@
+package view;
+
+public class InputView {
+    public void printing() {
```
1. 첫줄 개행 추가해주세요
2. printing은 명사형 이름이네요! 수정 해주세요~
3. print를 해주는 것은 outputView에서 해주는 것이 맞고 해당 InputView 클래스에선 값을 입력 받는 기능이 필요해보입니다!

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,43 @@
+
+    private static int i;
+    private static int result;
+    private static final int INITIAL_NUMBER = 0;
+    private static final int ARRAY_RESET = 100;
```
1. 상수를 제일 먼저 선언해주세요! 그리고 상수 선언 마치고 개행 한 칸 추가해주세요~
2. i, result를 static으로 선언한 이유가 있나요?
3. i는 선언할 필요가 없어보입니다!

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,43 @@
+    private static final int INITIAL_NUMBER = 0;
+    private static final int ARRAY_RESET = 100;
+
+    SplitString arr = new SplitString();
```
1. 접근제어자는 없어도 될까요?
2. arr란 네이밍은 적합하지 않아보입니다!

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,43 @@
+
+    public void plusCalculate() {
+        result = INITIAL_NUMBER;
+        arr.intArr = new int[ARRAY_RESET];
```
현재는 외부에서 값을 수정해주는데 객체에게 일을 시키는 건 어떨까요?

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,43 @@
+        result = INITIAL_NUMBER;
+        arr.intArr = new int[ARRAY_RESET];
+        for (i = INITIAL_NUMBER; i < arr.intArr.length; i++) {
+            result += arr.intArr[i];
```
1. divide인데 덧셈을 수행하네요! 수정해주세요
2. 나누는 수가 0일 때 문제가 생길 것 같은데 예외처리 해주세요!
3. 메서드에 접근제어자도 빠져있습니다!
4. 메서드명에서 실수를 줄이기 위해서 `calculate+연산명`과 같이 한다면 외부에서 해당 메서드들을 쓸 때 더 편할 것 같습니다.
5. result에 결과를 두지말고 반환하는 식으로 하는 건 어떨까요?

sosow0212 commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,34 @@
+
+class CalculatorTest {
+
+    @DisplayName("더하기 테스트")
+    @Test
+    void 더하기_테스트() {
```
이미 한글 메서드명인데 DisplayName은 없어도 될 것 같습니다 !

sosow0212 commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,34 @@
+    @DisplayName("더하기 테스트")
+    @Test
+    void 더하기_테스트() {
+        arithmeticCalculator cal = new arithmeticCalculator();
```
반복적으로 선언한다면 테스트 클래스 상단에 선언하는 건 어떨까요?

sosow0212 commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,34 @@
+    @Test
+    void 더하기_테스트() {
+        arithmeticCalculator cal = new arithmeticCalculator();
+        cal.plusCalculate(3, 5);
```
1. 결과 값 검증이 없는 것 같아요. 연산만 수행하는 것이 아닌 결과 값도 의도한대로 나오는지 변경 부탁드려요! 확인할 수 없는 구조라면 구조적인 문제는 아닌지 고민해보는 것도 좋을 것 같습니다!

2. given-when-then 패턴 공부해서 주석을 달며 사용해보면 어떨까요?

sosow0212 commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,34 @@
+        arithmeticCalculator cal = new arithmeticCalculator();
+        cal.multipleCaculate(3, 5);
+    }
+}
```
EOF가 발생하고 있네요! [해당 글](https://velog.io/@d-h-k/intellij-%ED%8C%8C%EC%9D%BC%EB%81%9D%EC%97%90-%EA%B0%9C%ED%96%89%EC%9D%84-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-%EC%B6%94%EA%B0%80%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95) 보시고 설정 부탁드려요

SEUNGBOONG replied:
클래스 앞에는 대문자로 해야하는데 아직 미숙했습니다. 다음 번엔 잘 지켜서 해보겠습니다.

SEUNGBOONG replied:
해당 글 공유 감사합니다!

sosow0212 commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,34 @@
+import org.junit.jupiter.api.Test;
+
+
+class CalculatorTest {
```
이 클래스는 무엇을 테스트 하는 것일까요?

ArithmeticCalculator, SplitString 별로 각각의 기능을 검증하는 테스트 클래스가 필요할 것 같습니다!

SEUNGBOONG replied:
단위 테스트로 쪼개서 해보겠습니다.

sosow0212 commented on src/test/java/domain/splitStringTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    @DisplayName("더하기 테스트")
```
더하기 테스트는 여기서도 하고 CalculatorTest에서도 진행하고 있네요
splitStringTest에서는 splitString 클래스 테스트만 하는 것이 좋아보입니다!

SEUNGBOONG replied:
아 CalculatrTest에서는 Step1 프로그램이고 
splitStringTest에서는 Step1_3 요구사항 추가했을때의 테스트입니다.

sosow0212 commented on src/test/java/domain/splitStringTest.java:
```diff
@@ -0,0 +1,62 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
```
구글 컨벤션에 의해 import에 와일드카드는 지양해주세요. 
[해당 글](https://invicr.github.io/2020/03/03/IntelliJ-Import-%EC%84%A4%EC%A0%95/) 보시고 설정 부탁드릴게요!

SEUNGBOONG replied:
글에서 설명하는 와일드카드가 * 이 맞을까요 ?

sosow0212 replied:
넵 `*`는 와일드카드 맞습니다!
`*` 이거는 해당 패키지 아래 있는 것을 모두 import한다라는 뜻인데, 그러지말고 위에 글 설정 따라하시고 묶여있는 import를 풀면 됩니다

sosow0212 commented on src/test/java/domain/splitStringTest.java:
```diff
@@ -0,0 +1,62 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class splitStringTest {
```
도메인 패키지에 있는 모든 테스트 클래스명이 소문자로 시작하고 있습니다.
SplitStringTest 와 같이 대문자로 시작해주세요!

SEUNGBOONG replied:
넵 !!

sosow0212 reviewed (changes requested):
수고하셨습니다 ~

자바를 시작한지 얼마 되지 않으셔서 컨벤션 부분에서 익숙치 않을텐데요, 적응하실 때까지는 컨벤션 위주로 리뷰를 달아보겠습니다! 

지금 기능구현이 제대로 되지 않은 것 같아요! 입력을 받고 출력을 잘 해주는지, 문제의 요구사항을 잘 지켰는지 확인해보고 다시 리뷰요청 주세요!! 고생하셨습니다 👍

SEUNGBOONG commented:
혹시 요구사항에는 계산기에 나눗셈을 하라는것이 나와있지 않은데 
만약 통합 테스트를 진행할 때 나눗셈은 어떻게 진행해야할까요 ?

SEUNGBOONG committed:
refactor :
 갈아 엎기 전 어떤 상태인지 알기 위한 커밋

SEUNGBOONG commented:
혹시 지금 타이밍에 람다식을 한번 공부하고 들어가는게 나을까요?

SEUNGBOONG committed:
refactor : 클린 코드를 최대한 지켜가며 수정해봤습니다.

SEUNGBOONG committed:
refactor : 처음에 성공한 테스트가 있었는데
도메인에 있던 SplitString.java 클래스를 view로 리팩토링하면서 실패한 코드로 바뀌었습니다.

SEUNGBOONG committed:
refactor : MVC 모델 구현후
최종적으로 Controller에서 합치려고했습니다.

SEUNGBOONG committed:
refactor : 이부분은 리팩토링하면서 통합테스트를 구현해보려고했는데, 어려움이 있었습니다. 될떄가 있고 안될때가있어 TDD를 지키지못한 것같아 질문을 드리고싶은 부분입니다.

SEUNGBOONG committed:
refactor : 이부분은 리팩토링하면서 통합테스트를 구현해보려고했는데, 어려움이 있었습니다. 될떄가 있고 안될때가있어 TDD를 지키지못한 것같아 질문을 드리고싶은 부분입니다.

SEUNGBOONG committed:
refactor : InputView 에서 스캐너를 사용함으로써 마지막 Controller에서 최종적으로 사용하기위해 구현해봤습니다.

SEUNGBOONG committed:
refactor : 가장 많은 리팩토링 부분입니다. 결국 Controller을 완성하는 부분에서 의구심이 들었습니다. 도메인은 뷰를 의존해선 안된다는 것이 떠올라 여태 했던것들을 보면 도메인에서 뷰를 의존하게 되는 것 같습니다. 그래서 도메인 코드를 Outputview 로 리팩토링했습니다.

SEUNGBOONG committed:
refactor : 이부분은 원래 도메인에서 구현했던 부분입니다. 결국 처음 설계할 때 도메인이라고 생각한 부분이 착오 인 것 같습니다. 처음 설계한대로 하게되면
domain에서 view를 의존해버립니다. 그래서 큰맘먹고 MVC 패턴을 위한 리팩토링을 진행했습니다.

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,51 @@
+
+public class ArithmeticCalculator {
+
+    private static final int INITIAL_NUMBER = 0;
+    private static final int MULTIPLE_DIVIDE_NUMBER = 1;
+    private int[] intArr;
+    private int result = INITIAL_NUMBER;
```
```suggestion
    private static final int INITIAL_NUMBER = 0;
    private static final int MULTIPLE_DIVIDE_NUMBER = 1;
    
    private int[] intArr;
    private int result = INITIAL_NUMBER;
```

sosow0212 replied:
1. intArr의 네이밍이 조금 더 구체적이면 좋겠습니다.
2. 배열과 List의 차이는 무엇일까요?

SEUNGBOONG replied:
헐 가변 불변 차이인것 같습니다. 
갑자기 잘못헀다는게 떠올랐습니다. 
가변적으로 하는게 좋다는 판단이 들었습니다.

SEUNGBOONG replied:
Array 로 구현한것이 조회할 때 더 빠르다고 합니다!

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,51 @@
+    private int[] intArr;
+    private int result = INITIAL_NUMBER;
+
+    public int getResult() {
+        return result;
+    }
```
result 변수가 있을 필요가 있을까요??

calculate 메서드에서 값을 반환해주는데 getResult()는 필요가 없어보이네요 ㅎㅎ
그리고 계산기 기능도 전역변수 result 말고 지역변수로 관리하는 건 어떨까요?

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,51 @@
+        return result;
+    }
+
+    public int plusCalculate(String str) {
+        resetArr(str);
+        for (int i = INITIAL_NUMBER; i < intArr.length; i++) {
+            result += intArr[i];
+        }
+        return result;
+    }
```
1. 지금 구조면 resetArr에 str이 넘어가서 배열의 자리를 결정한 후 아무 값도 넣지 않아서 result에 아무 값도 담기지 않을 거 같아요! 다른 부분도 마찬가지인데 확인해주세요~
2. 매개변수 str도 뜻을 주면 더 좋을 것 같습니다

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,35 @@
+package view;
+
+public class OutputView {
+    private final String DISTINCTION_FRONT = "//";
```
1. 첫줄 개행 추가해주세요!
2. 상수인데 static final 로 바꿔주세요~

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,35 @@
+    private final int START_NUM = 2;
+    private final int NEXT_START_NUM = 1;
+
+    private int i;
+    private int[] intArr;
```
변수가 어떤 걸 의미하는지 알기 힘들 수 있을 것 같아요~ 조금 더 풀어서 쓰면 좋을 것 같습니다!

intelliJ -> `fn + shift + F6` 단축키 활용해서 바꿔보면 좋을 것 같습니다

sosow0212 commented on src/test/java/controller/CalculatorControllerTest.java:
```diff
@@ -0,0 +1,26 @@
+import org.junit.jupiter.api.Test;
+import view.InputView;
+import view.OutputView;
+
+
```
불필요한 개행이 있네요!

sosow0212 commented on src/test/java/controller/CalculatorControllerTest.java:
```diff
@@ -0,0 +1,26 @@
+
+
+
+public class CalculatorControllerTest {
```
여기서 해당 통합테스트는 의미가 없을 것 같아요
해당 클래스 제거하고 단위테스트만 해도 될 것 같습니다!

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,35 @@
+package view;
+
+public class OutputView {
```
OutputView인데 문자열 파싱하는 로직이 섞여있는 것 같아요
문자열 분리는 다른 클래스에서 진행하고 OutputView는 진짜 넘어온 값을 출력만 하도록 바꾸면 좋을 것 같습니다!

sosow0212 reviewed (changes requested):
뷰에 도메인 로직이 섞여있는 것 같아요 ㅠㅠ
해당 미션에서는 테스트도 통합보다는 도메인 단위 테스트만 확실하게 진행해도 될 것 같습니다!
한 번 더 RC하겠습니다 🙇🏻‍♂️

SEUNGBOONG committed:
refactor :
배열을 사용하다가, 리스트로 변환해서 진행했습니다.

SEUNGBOONG committed:
refactor :
OutputView를 수정 한 후 테스트를 진행했습니다.

SEUNGBOONG committed:
refactor :
예외 처리 테스트도 성공적으로 끝냈습니다.

SEUNGBOONG committed:
refactor :
리스트 사용하던 것을
배열로 바꿔보았습니다.

SEUNGBOONG committed:
refactor :
바꾼 배열로 테스트 통과했습니다.

SEUNGBOONG committed:
refactor :
- 응집도를 높이기 위한 메서드 분리
- static 지양을 위한 수정
- 캡슐화를 위한 public->private 수정

SEUNGBOONG committed:
refactor :
- 응집도를 높이기 위한 메서드 분리
- static 지양을 위한 수정
- 캡슐화를 위한 public->private 수정

SEUNGBOONG committed:
refactor :
- 응집도를 높이기 위한 메서드 분리
- static 지양을 위한 수정
- 캡슐화를 위한 public->private 수정
private 수정 후 get으로 입력값 받기 완료

SEUNGBOONG committed:
refactor :
테스트 통과

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,63 @@
+        this.result = INITIAL_RESULT_NUM;
+    }
+
+    private int plusCalculate() {
```
해당 메서드를 private로 감춰두면 테스트에서 검증하기 힘들고, 사용자도 쓸 수 없을 것 같아요!
나중에 계산부는 실제 사용하니 public으로 바꿔주세요~

SEUNGBOONG replied:
넵 수정 완료했습니다.

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,63 @@
+    private static final int INITIAL_NUMBER = 1;
+
+    private final ArrayList<Integer> numbers;
+    private int result;
```
numbers에 입력한 숫자들이 들어가고 지금 보면 뺄셈, 나눗셈 같은 경우는 첫 수 기준으로 진행하는 거 같은데
현재 구조라면 result 필드는 필요 없을 것 같아요~

numbers에 생성자로 이미 값이 생성된 상태라면 계산 결과를 각 로직에 맞게 수행한 후 결과 값을 반환하는게 테스트할 때에도 더 편하고 불필요한 필드도 제거할 수 있을 것 같습니다!

SEUNGBOONG replied:
수정 완료했습니다. 무조건 필드에 선언하고 스타트하는것이 꼭 좋은 것 만은 아니군요!

sosow0212 commented on src/main/java/domain/SplitExceptionCal.java:
```diff
@@ -0,0 +1,36 @@
+        parsedValue = parseInput(input);
+    }
+
+    //"//;\n1;2;3"
```
불필요한 주석은 제거해주세요!

SEUNGBOONG replied:
넵!

sosow0212 commented on src/main/java/domain/SplitExceptionCal.java:
```diff
@@ -0,0 +1,36 @@
+    private List<String> parsedValue;
+
+    private SplitExceptionCal(String input) {
+        parsedValue = parseInput(input);
```
생성자에서 모든 로직을 다 처리해주네요~ getter가 있어야 파싱된 숫자를 꺼낼 수 있을 것 같습니다!

sosow0212 commented on src/main/java/domain/SplitExceptionCal.java:
```diff
@@ -0,0 +1,36 @@
+        return Arrays.asList(values);
+    }
+
+    private void parseException(String value) {
```
사용하지 않는 메서드 같은데 지워주세요!

SEUNGBOONG replied:
넵!

sosow0212 commented on src/main/java/domain/SplitExceptionCal.java:
```diff
@@ -0,0 +1,36 @@
+            throw new RuntimeException(e);
+        }
+    }
+
```
불필요한 공백 제거해주세요~

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,31 @@
+    private ArrayList<Integer> parsedValues = new ArrayList<>();
+
+    public List<Integer> parseString(String str) {
+
```
요기도 불필요한 공백이 존재하네요!

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,31 @@
+    private static final String DISTINCTION_BACK = "\n";
+    private static final int FOR_DISTINCTION_NUM = 1;
+
+    private ArrayList<Integer> parsedValues = new ArrayList<>();
```
OutputView에서 parsedValues가 필요할까요?
OutputView에서는 요구에 따른 가공된 결과 값만 매개변수로 넘겨서 출력만 담당해도 될 것 같아요

현재는 OutputView에서 입력값을 파싱하는데, 이건 뷰의 역할보다는 도메인의 역할로 빼는 것이 좋을 것 같습니다!

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,31 @@
+                try {
+                    int parsedNum = Integer.parseInt(num.trim());
+                    parsedValues.add(parsedNum);
+                } catch (NumberFormatException e) {
```
catch를 할 때 Exception은 발생하지 않아도 될까요?
Exception에 대해서 공부해보면 좋을 것 같습니다!

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,31 @@
+
+        if (str.startsWith(DISTINCTION_FRONT) && str.contains(DISTINCTION_BACK)) {
+            String separator = str.substring(DISTINCTION_FRONT.length(), DISTINCTION_FRONT.length() + FOR_DISTINCTION_NUM);
+            String numbersList = str.substring(str.indexOf(DISTINCTION_BACK) + FOR_DISTINCTION_NUM);
+            String[] numbersArray = numbersList.split(separator);
```
변수명에 자료형의 이름을 쓰는 것보다 적절한 변수명을 사용하면 좋을 것 같아요~

String numbersList라면 List 자료형일 거 같은데 String인 점도 조금 모호하고, 파싱한 값이라면 더 적절한 값으로 변경해주세요!

sosow0212 commented on src/main/java/controller/CalculatorController.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    void startCalculator(String str) {
+        ArithmeticCalculator calculator = new ArithmeticCalculator();
+        String input = inputView.calculatorInput();
+        calculator.plusCalculate(outputView.splitStr(input));
```
1. ArithmeticCaculator에 생성자 값이 들어가지 않아서 실행이 안 돼요 ㅠ
2. outputView.splitStr(...) 메서드가 존재하지 않아서 실행이 안 됩니다!

sosow0212 commented on src/test/java/domain/ArithmeticCalculatorTest.java:
```diff
@@ -0,0 +1,74 @@
+    @Test
+    void 덧셈을_한다() {
+        // given
+        ArrayList<Integer> numbers = new ArrayList<>();
+        ArithmeticCalculator calculator = new ArithmeticCalculator(numbers);
+        numbers.add(3);
+        numbers.add(5);
+        numbers.add(7);
```
numbers에 생성할 때 값을 넣어주고
ArithmeticCalculator로 numbers를 넣어준 후 테스트를 작동하는 건 어떨까요?

지금 결과는 나오지만, 
현재 구조는 numbers의 빈 값을 바탕으로 ArithmeticCalculator를 만들고 numbers에 add를 해서 실행 및 검증하고 있습니다.

미리 준비된 값을 바탕으로 도메인에 넣어준 후 검증하는 것이 좋을 것 같아요.

`여기서 질문`
- numbers.add()가 되기 전에 ArithmeticCalculator에 아무 값도 없는 numbers가 들어가는데 결과가 잘 나오는 이유는 무엇일까요? (자바의 프리미티브 타입과 레퍼런스 타입에 대해서 공부해보면 좋을 것 같아요 {주소값 위주로})

sosow0212 commented on src/test/java/domain/ArithmeticCalculatorTest.java:
```diff
@@ -0,0 +1,74 @@
+            assertEquals(2, result);
+        }
+    }
+}
```
EOF 문제가 발생합니다!

sosow0212 commented on src/test/java/domain/splitStringTest.java:
```diff
@@ -0,0 +1,37 @@
+//package domain;
```
전부 주석한 해당 클래스는 지워주세요

sosow0212 commented on src/test/java/domain/ArithmeticCalculatorTest.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @DisplayName("나눗셈 테스트")
+    @Nested
```
여기에 예외도 검증해보면 좋을 것 같습니다!

sosow0212 reviewed (changes requested):
처음에 비해 많이 구조화되고 깔끔해진게 보이네요!
다만 제출 전에 작동을 해보면 좋을 것 같아요 ㅠ 현재는 실행되지 않고 있습니다.
몇 부분만 고쳐주시면 좋을 것 같아요.

아마 마지막 리뷰가 될 수 있을 것 같네요! 수고하셨습니다 👍

SEUNGBOONG committed:
Fix :

SEUNGBOONG committed:
Fix :
SPLIT_WORD 추가

SEUNGBOONG committed:
Refactor :
- 최대한 클린 코드를 지키며 수정했습니다.

SEUNGBOONG committed:
Refactor :
- 계속해서 실행이 안됐습니다 .
그래서 --stack force 확인 결과 parseInput 부분에서 배열 입력을 제대로 받지 못한 상태였습니다. 그래서 수정했습니다.

SEUNGBOONG committed:
Refactor :
- 인풋뷰 간단하게 수정

SEUNGBOONG committed:
Test :
- 테스트 작성완료

SEUNGBOONG committed:
Refactor :
-  오류 부분확인후 메서드 수정했습니다. 객체를 잘 나누니까
수정할때 편하다는 것을 깨달았습니다.

SEUNGBOONG committed:
fix:
- 버그 수정 중 바뀐 것들

SEUNGBOONG committed:
Refactor :
- OutputView에서는 출력값만을 다루게 해서 리팩했습니다.

SEUNGBOONG committed:
Refactor :
- OutputView에서는 출력값만을 다루게 해서 리팩했습니다!!

SEUNGBOONG committed:
Fix: 문구 삭제

SEUNGBOONG committed:
Fix: minusCalculate() 메서드 람다식으로 변환

sosow0212 commented on src/main/java/domain/ArithmeticCalculator.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public int minusCalculate() {
+        return numbers.stream().reduce(INITIAL_RESULT_NUM,
```
오 스트림까지~~
객체지향 생활체조 4번 따르면 개행 이렇게 바뀌겠네요!

```suggestion
        return numbers.stream()
                       .reduce(INITIAL_RESULT_NUM,(result, number) -> result - number);
```

sosow0212 commented on src/main/java/domain/SplitExceptionCal.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.List;
+
+public class SplitExceptionCal {
+    private static final int CUT_NUMBER = 5;
```
여기도 개행 하나 빠졌습니다!

sosow0212 commented on src/main/java/domain/SplitExceptionCal.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class SplitExceptionCal {
```
String을 파싱해주는 거니 네이밍을 바꾸면 좋을 것 같긴 합니다~

sosow0212 reviewed (approved):
처음 제출에 비해 코드가 많이 깔끔해진 것 같아서 아주 좋습니다 👍
매일 도서관 가서 공부하신게 효과가 있네요!

이번 미션에서는 구조도 구조지만 컨벤션 위주로 리뷰를 했습니다.
다음 미션 때에도 컨벤션 잘 지켜주시고 구조적인 부분도 한 번 신경써봐요!!

단톡방에 올린 키워드도 잘 학습하셔서 다음 미션 때 더 발전한 모습 봤으면 좋겠습니다.
수고하셨습니다 👏
